rust/connlib/libs/tunnel/src/tun_android.rs

```diff
@@ -152,8 +171,10 @@ impl IfaceConfig {
         &mut self,
         route: IpNetwork,
         callbacks: &CallbackErrorFacade<impl Callbacks>,
-    ) -> Result<()> {
-        callbacks.on_add_route(route)
+    ) -> Result<RawFd> {
+        let result = callbacks.on_add_route(route)?;
+        self.0.replace_fd(result);
```

Comment:
Got stuck here 🤠

Reply:
```
error[E0596]: cannot borrow data in an `Arc` as mutable
   --> connlib/libs/tunnel/src/tun_android.rs:176:9
    |
176 |         self.0.replace_fd(result);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Arc<IfaceDevice>`

For more information about this error, try `rustc --explain E0596`.
error: could not compile `firezone-tunnel` (lib) due to previous error
```

Reply:
The problem here is that `.0` is an `Arc` which is just a shared pointer and does not allow for mutable access. You'd have to wrap the `IfaceDevice` in a `Mutex` to call a function with a mutable self receiver (`&mut self`) on it.

I may not have the full context but I wonder why this is wrapped in an `Arc` in the first place? `IfaceConfig` does not implement `Clone` so you should be able to drop the wrapping in `Arc` and thus access the inner field with a regular mutable reference.

Reply:
`IFaceConfig` is just a wrapper around an `IFaceDevice` which is actually accessed from multiple tasks that's why it's wrapped in an `Arc`.

We shouldn't wrap it with a `Mutex` or even a `RWLock` since I think the performance hit would be too bad for something that is in the hot path.

The way to go will be to drop all `IFaceConfig` and `IFaceDevice` related tasks and restart them with the new `fd` after draining the old one.

Reply:
> We shouldn't wrap it with a `Mutex` or even a `RWLock` since I think the performance hit would be too bad for something that is in the hot path.

How often would a write-lock actually be taken? Configuration changes aren't that common, are they?

> The way to go will be to drop all `IFaceConfig` and `IFaceDevice` related tasks and restart them with the new `fd` after draining the old one.

This works too but I am not familiar enough with the code to say whether or not this creates complexity. In a way, it is the cleaner solution because mutable state is usually more difficult to understand.

Reply:
> How often would a write-lock actually be taken? Configuration changes aren't that common, are they?

The problem is that `IFaceConfig` wraps `IFaceDevice`, which is not only used for configuration but also to read/write packages. 

Even if we refactored and separated this, when we update the `fd` we want to update the reference that is held by `IFaceDevice` since what's updated here is the file descriptor used to read/write packets.